ITest .getTestName() doesn't return the actual test name for skipped methods

### TestNG Version
> Note: only the latest version is supported

6.14.3

### Expected behavior

If a test class has implemented `org.testng.ITest` then TestNG should be invoking `getTestName()` for skipped tests.

### Actual behavior

Currently TestNG ends up retrieving just the actual method name instead of invoking `getTestName()` for skipped tests.

### Is the issue reproductible on runner?

- [X] Shell
- [X] Maven
- [X] Gradle
- [ ] Ant
- [ ] Eclipse
- [ ] IntelliJ
- [ ] NetBeans


### Test case sample
> Please, share the test case (as small as possible) which shows the issue

Below snipped can be used to reproduce the problem.

**Test class**

```java
import org.testng.ITest;
import org.testng.annotations.Listeners;
import org.testng.annotations.Test;

import java.util.UUID;

@Listeners(CustomTestNGReporter.class)
public class BaseTest implements ITest {
    private static String uuid = UUID.randomUUID().toString();

    public static String getUuid() {
        return uuid;
    }

    private String uri;

    public BaseTest() {
        this.uri = uuid;
        System.err.println("uri at the time of object creation " + uri);
    }


    @Test
    public void test1() {
        throw new RuntimeException("failed");
    }


    @Test(dependsOnMethods = "test1")
    public void test2() { }

    public String getTestName() {
        return uri;
    }
}
```

**Listener**
```java
import org.testng.IReporter;
import org.testng.IResultMap;
import org.testng.ISuite;
import org.testng.ITestContext;
import org.testng.xml.XmlSuite;

import java.util.LinkedList;
import java.util.List;

public class CustomTestNGReporter implements IReporter {
    private List<String> logs = new LinkedList<>();

    public List<String> getLogs() {
        return logs;
    }

    public void generateReport(List<XmlSuite> xmlSuites, List<ISuite> suites, String outputDirectory) {
        getTestMehodSummary(suites);
    }

    private void getTestMehodSummary(List<ISuite> suites) {
        suites.forEach(iSuite -> iSuite.getResults().values().forEach(each -> {
            ITestContext testObj = each.getTestContext();
            System.err.println("Getting failed test method related data");
            getTestMethodReport(testObj.getFailedTests());
            System.err.println("Getting skipped test method related data");
            getTestMethodReport(testObj.getSkippedTests());
            System.err.println("Getting passed test method related data");
            getTestMethodReport(testObj.getPassedTests());

        }));
    }

    private void getTestMethodReport(IResultMap testResultMap) {
        testResultMap.getAllResults().forEach(iTestResult -> logs.add(iTestResult.getName()));
    }
}
```

**Test runner class**
```java
import org.assertj.core.api.Assertions;
import org.testng.ITestNGListener;
import org.testng.TestNG;
import org.testng.annotations.Test;

public class TestRunner {
    @Test
    public void testMethod() {
        TestNG testng = new TestNG();
        testng.setTestClasses(new Class[] {BaseTest.class});
        CustomTestNGReporter reporter = new CustomTestNGReporter();
        testng.addListener((ITestNGListener) reporter);
        testng.run();
        System.err.println(reporter.getLogs());
        Assertions.assertThat(reporter.getLogs()).containsExactly(BaseTest.getUuid(), BaseTest.getUuid());
    }
}
```
This test currently fails with the below output:

```
uri at the time of object creation 10871d39-1691-47b8-905f-5d244aa99f7a

===============================================
Command line suite
Total tests run: 2, Failures: 1, Skips: 1
===============================================

Getting failed test method related data
Getting skipped test method related data
Getting passed test method related data
[10871d39-1691-47b8-905f-5d244aa99f7a, test2]

java.lang.AssertionError: 
Expecting:
  <["10871d39-1691-47b8-905f-5d244aa99f7a", "test2"]>
to contain exactly (and in same order):
  <["10871d39-1691-47b8-905f-5d244aa99f7a",
    "10871d39-1691-47b8-905f-5d244aa99f7a"]>
but some elements were not found:
  <[]>
and others were not expected:
  <["test2"]>


	at com.rationaleemotions.stackoverflow.qn49693956.TestRunner.testMethod(TestRunner.java:17)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:583)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:719)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:989)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
	at org.testng.TestRunner.privateRun(TestRunner.java:648)
	at org.testng.TestRunner.run(TestRunner.java:505)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:455)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:450)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:415)
	at org.testng.SuiteRunner.run(SuiteRunner.java:364)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:84)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1208)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1137)
	at org.testng.TestNG.runSuites(TestNG.java:1049)
	at org.testng.TestNG.run(TestNG.java:1017)
	at org.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:72)
	at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:123)


===============================================
Default Suite
Total tests run: 1, Failures: 1, Skips: 0
===============================================
```

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,4 +1,5 @@
 Current
+Fixed: GITHUB-1756: ITest .getTestName() doesn't return the actual test name for skipped methods (Krishnan Mahadevan)
 Fixed: GITHUB-1602: beforeInvocation method don't get called for skipped tests (Krishnan Mahadevan)
 Fixed: GITHUB-549 and GITHUB-780: Introduce onlyForGroups attribute for @BeforeMethod and @AfterMethod (Sergei Tachenov)
 Fixed: GITHUB-1745: Support native injection for @Factory methods (Krishnan Mahadevan)

--- a/src/main/java/org/testng/internal/Invoker.java
+++ b/src/main/java/org/testng/internal/Invoker.java
@@ -906,7 +906,7 @@ public class Invoker implements IInvoker {
       //
       // Not okToProceed. Test is being skipped
       //
-      ITestResult result = registerSkippedTestResult(testMethod, null, System.currentTimeMillis(),
+      ITestResult result = registerSkippedTestResult(testMethod, instance, System.currentTimeMillis(),
           new Throwable(okToProceed));
       m_notifier.addSkippedTest(testMethod, result);
       return Collections.singletonList(result);

--- a/src/test/java/test/dependent/DependentTest.java
+++ b/src/test/java/test/dependent/DependentTest.java
@@ -102,7 +102,7 @@ public class DependentTest extends BaseTest {
     verifyInstanceNames("Failed", getFailedTests(),
         new String[]{"f#2"});
     verifyInstanceNames("Skipped", getSkippedTests(),
-        new String[]{"g#"});
+        new String[]{"g#2"});
   }
 
   @Test

--- a/src/test/java/test/reports/ReportTest.java
+++ b/src/test/java/test/reports/ReportTest.java
@@ -7,13 +7,13 @@ import org.testng.reporters.FailedReporter;
 import org.testng.reporters.TestHTMLReporter;
 import org.testng.xml.Parser;
 import org.testng.xml.XmlSuite;
-import org.xml.sax.SAXException;
 import test.InvokedMethodNameListener;
 import test.SimpleBaseTest;
 import test.TestHelper;
+import test.reports.issue1756.CustomTestNGReporter;
+import test.reports.issue1756.SampleTestClass;
 import test.simple.SimpleSample;
 
-import javax.xml.parsers.ParserConfigurationException;
 import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -33,7 +33,7 @@ public class ReportTest extends SimpleBaseTest {
     XmlSuite suite = createXmlSuite(suiteName, testName, SimpleSample.class);
 
     TestNG tng = create(outputDir, suite);
-    tng.addListener((ITestNGListener) new TestHTMLReporter());
+    tng.addListener(new TestHTMLReporter());
 
     Path f = getHtmlReportFile(outputDir, suiteName, testName);
 
@@ -54,7 +54,7 @@ public class ReportTest extends SimpleBaseTest {
     createXmlTest(xmlSuite, testName2);
 
     TestNG tng = create(outputDirectory, xmlSuite);
-    tng.addListener((ITestNGListener) new TestHTMLReporter());
+    tng.addListener(new TestHTMLReporter());
 
     Path f = getHtmlReportFile(outputDirectory, suiteName, testName);
     Path f2 = getHtmlReportFile(outputDirectory, suiteName, testName2);
@@ -76,7 +76,7 @@ public class ReportTest extends SimpleBaseTest {
     XmlSuite xmlSuiteB = createXmlSuite(suiteNameB, testName, SampleB.class);
 
     TestNG tng = create(outputDirectory, xmlSuiteA, xmlSuiteB);
-    tng.addListener((ITestNGListener) new TestHTMLReporter());
+    tng.addListener(new TestHTMLReporter());
 
     Path f1 = getHtmlReportFile(outputDirectory, suiteNameA, testName);
     Path f2 = getHtmlReportFile(outputDirectory, suiteNameB, testName);
@@ -98,7 +98,7 @@ public class ReportTest extends SimpleBaseTest {
     Path outputDirectory = TestHelper.createRandomDirectory();
 
     TestNG tng = create(outputDirectory, SampleA.class, SampleB.class);
-    tng.addListener((ITestNGListener) new TestHTMLReporter());
+    tng.addListener(new TestHTMLReporter());
 
     Path fileA = outputDirectory.resolve("SuiteA-JDK5");
     Path fileB = outputDirectory.resolve("SuiteB-JDK5");
@@ -126,7 +126,7 @@ public class ReportTest extends SimpleBaseTest {
         ReportTest.m_success = (output != null && output.size() > 0);
       }
     };
-    tng.addListener((ITestNGListener) listener);
+    tng.addListener(listener);
     tng.run();
 
     Assert.assertTrue(m_success);
@@ -136,7 +136,7 @@ public class ReportTest extends SimpleBaseTest {
   public void reportLogShouldBeAvailableWithListener() {
     TestNG tng = create(ListenerReporterSample.class);
     InvokedMethodNameListener listener = new InvokedMethodNameListener();
-    tng.addListener((ITestNGListener) listener);
+    tng.addListener(listener);
 
     Reporter.clear();
 
@@ -152,13 +152,13 @@ public class ReportTest extends SimpleBaseTest {
   public void github1090() {
     TestNG tng = create(GitHub447Sample.class);
     GitHub447Listener reporter = new GitHub447Listener();
-    tng.addListener((ITestNGListener) reporter);
+    tng.addListener(reporter);
     tng.run();
 
     List<Object[]> parameters = reporter.getParameters();
     Assert.assertEquals(parameters.size(), 5);
     Assert.assertEquals(parameters.get(0)[0].toString(), "[]");
-    Assert.assertEquals(parameters.get(0)[1], null);
+    Assert.assertNull(parameters.get(0)[1]);
     Assert.assertEquals(parameters.get(0)[2].toString(), "[null]");
     Assert.assertEquals(parameters.get(1)[0].toString(), "[null]");
     Assert.assertEquals(parameters.get(1)[1], "dup");
@@ -170,7 +170,7 @@ public class ReportTest extends SimpleBaseTest {
     Assert.assertEquals(parameters.get(3)[1], "str");
     Assert.assertEquals(parameters.get(3)[2].toString(), "[null, dup, dup, str]");
     Assert.assertEquals(parameters.get(4)[0].toString(), "[null, dup, dup, str]");
-    Assert.assertEquals(parameters.get(4)[1], null);
+    Assert.assertNull(parameters.get(4)[1]);
     Assert.assertEquals(parameters.get(4)[2].toString(), "[null, dup, dup, str, null]");
   }
 
@@ -183,14 +183,14 @@ public class ReportTest extends SimpleBaseTest {
   }
 
   @Test(dataProvider = "dp")
-  public void runFailedTestTwiceShouldBeConsistent(Class<?> testClass, String[] succeedMethods, String[] failedMethods) throws IOException, ParserConfigurationException, SAXException {
+  public void runFailedTestTwiceShouldBeConsistent(Class<?> testClass, String[] succeedMethods, String[] failedMethods) throws IOException {
     Path outputDirectory = TestHelper.createRandomDirectory();
 
     TestNG tng = create(outputDirectory, testClass);
 
     InvokedMethodNameListener listener = new InvokedMethodNameListener();
-    tng.addListener((ITestNGListener) listener);
-    tng.addListener((ITestNGListener) new FailedReporter());
+    tng.addListener(listener);
+    tng.addListener(new FailedReporter());
 
     tng.run();
 
@@ -206,14 +206,23 @@ public class ReportTest extends SimpleBaseTest {
     }
   }
 
-  private static Path checkFailed(Path testngFailedXml, String... failedMethods) throws IOException, ParserConfigurationException, SAXException {
+  @Test(description = "GITHUB-1756")
+  public void testToEnsureSkippedTestsHaveProviderITestNameRetrieved() {
+    TestNG testng = create(SampleTestClass.class);
+    CustomTestNGReporter reporter = new CustomTestNGReporter();
+    testng.addListener(reporter);
+    testng.run();
+    assertThat(reporter.getLogs()).containsExactly(SampleTestClass.getUuid(), SampleTestClass.getUuid());
+  }
+
+  private static Path checkFailed(Path testngFailedXml, String... failedMethods) throws IOException {
     Path outputDirectory = TestHelper.createRandomDirectory();
 
     List<XmlSuite> suites = new Parser(Files.newInputStream(testngFailedXml)).parseToList();
     TestNG tng = create(outputDirectory, suites);
     InvokedMethodNameListener listener = new InvokedMethodNameListener();
-    tng.addListener((ITestNGListener) listener);
-    tng.addListener((ITestNGListener) new FailedReporter());
+    tng.addListener(listener);
+    tng.addListener(new FailedReporter());
 
     tng.run();
 

--- a/src/test/java/test/reports/issue1756/CustomTestNGReporter.java
+++ b/src/test/java/test/reports/issue1756/CustomTestNGReporter.java
@@ -1,0 +1,36 @@
+package test.reports.issue1756;
+
+import org.testng.IReporter;
+import org.testng.IResultMap;
+import org.testng.ISuite;
+import org.testng.ITestContext;
+import org.testng.xml.XmlSuite;
+
+import java.util.LinkedList;
+import java.util.List;
+
+public class CustomTestNGReporter implements IReporter {
+    private List<String> logs = new LinkedList<>();
+
+    public List<String> getLogs() {
+        return logs;
+    }
+
+    public void generateReport(List<XmlSuite> xmlSuites, List<ISuite> suites, String outputDirectory) {
+        getTestMehodSummary(suites);
+    }
+
+    private void getTestMehodSummary(List<ISuite> suites) {
+        suites.forEach(iSuite -> iSuite.getResults().values().forEach(each -> {
+            ITestContext testObj = each.getTestContext();
+            getTestMethodReport(testObj.getFailedTests());
+            getTestMethodReport(testObj.getSkippedTests());
+            getTestMethodReport(testObj.getPassedTests());
+
+        }));
+    }
+
+    private void getTestMethodReport(IResultMap testResultMap) {
+        testResultMap.getAllResults().forEach(iTestResult -> logs.add(iTestResult.getName()));
+    }
+}

--- a/src/test/java/test/reports/issue1756/SampleTestClass.java
+++ b/src/test/java/test/reports/issue1756/SampleTestClass.java
@@ -1,0 +1,34 @@
+package test.reports.issue1756;
+
+import org.testng.ITest;
+import org.testng.annotations.Listeners;
+import org.testng.annotations.Test;
+
+import java.util.UUID;
+
+@Listeners(CustomTestNGReporter.class)
+public class SampleTestClass implements ITest {
+    private static String uuid = UUID.randomUUID().toString();
+
+    public static String getUuid() {
+        return uuid;
+    }
+
+    private String uri;
+
+    public SampleTestClass() {
+        this.uri = uuid;
+    }
+
+    @Test
+    public void test1() {
+        throw new RuntimeException("failed");
+    }
+
+    @Test(dependsOnMethods = "test1")
+    public void test2() { }
+
+    public String getTestName() {
+        return uri;
+    }
+}

